Spec edits for incremental delivery, Response Section and Examples Appendix

robrichard opened the pull request:
Extracted from the full PR (https://github.com/graphql/graphql-spec/pull/1110) and targeting an integration branch to aid in review.

The response section is especially important, as it defines all of the terms which will be used in the execution algorithm

Helpful reference material:

* Response format examples: https://github.com/graphql/defer-stream-wg/discussions/69
* Glossary: https://github.com/graphql/defer-stream-wg/discussions/106#discussion-7942514
* GraphQL Conf talk: https://www.youtube.com/watch?v=LEyDeNoobT0

robrichard committed:
Spec edits for incremental delivery, Examples

robrichard committed:
Spec edits for incremental delivery, Section 7 only

BoD commented on spec/Appendix E -- Examples.md:
````diff
@@ -0,0 +1,191 @@
+}
+```
+
+Subsequent response 2, contains the remaining data from the
````
```suggestion
Subsequent response 2 contains the remaining data from the
```

BoD commented on spec/Section 7 -- Response.md:
```diff
@@ -323,15 +388,171 @@ discouraged.
+
+#### Incremental List Result
+
+:: An _incremental list result_ is a _incremental result_ used to deliver
```
```suggestion
:: An _incremental list result_ is an _incremental result_ used to deliver
```

BoD commented on spec/Section 7 -- Response.md:
```diff
@@ -323,15 +388,171 @@ discouraged.
+
+#### Incremental Object Result
+
+:: An _incremental object result_ is a _incremental result_ used to deliver
```
```suggestion
:: An _incremental object result_ is an _incremental result_ used to deliver
```

robrichard committed:
fix typos

benjie reviewed (dismissed):
I don't know if it's just me, but I feel like the examples could use more narrative and explanation that this is just one way the query could turn out. Here's how I'd change the beginning of the first example:

robrichard committed:
Update to address feedback in exampled appendix

robrichard committed:
Update terminology

Initial Execution Result > Initial Incremental Stream Result
Execution Update Result > Incremental Stream Update Result

robrichard committed:
allow hasNext: false on initial result

robrichard committed:
pending is must on initial stream result

robrichard committed:
PR feedback

robrichard committed:
value => payload when describing what is yielded by a stream

robrichard committed:
Update description of _initial incremental stream result_ & _incremental stream update result_

robrichard committed:
fix incorrect term

benjie commented on spec/Appendix E -- Examples.md:
```diff
@@ -0,0 +1,210 @@
+- a {"hasNext"} entry with the value {true}, indicating that the response is not
+  yet complete.
+
+If an error were to occur, it would also have an {"error"} entry; but not in
```
```suggestion
If an error were to occur, it would also have an {"errors"} entry; but not in
```

benjie commented on spec/Appendix E -- Examples.md:
```diff
@@ -0,0 +1,210 @@
+Depending on the behavior of the backend and the time at which the deferred and
+streamed resources resolve, the stream may produce results in different orders.
+In this example, our first _incremental stream update result_ contains the
+deferred data and the first streamed list item. There is one _completed result_,
```
Maybe change "completed result" to "incremental completion result" to tie it in with the "incremental" naming? (Also to be more specific.)

```suggestion
deferred data and the first streamed list item. There is one _incremental completion result_,
```

robrichard replied:
Updated now to _incremental completion notice_

benjie commented on spec/Appendix E -- Examples.md:
```diff
@@ -0,0 +1,210 @@
+In this example, the underlying iterator does not close synchronously so
+{"hasNext"} is set to {true}. If this iterator did close synchronously,
+{"hasNext"} would be set to {false} and this would be the final incremental
+stream update result.
```
I think "could" rather than "would" to allow vendor flexibility?

```suggestion
{"hasNext"} is set to {true}. If this iterator did close synchronously,
{"hasNext"} could be set to {false} and make this the final incremental
stream update result.
```

benjie commented on spec/Appendix E -- Examples.md:
```diff
@@ -0,0 +1,210 @@
+The third _incremental stream update result_ contains no incremental data.
+{"hasNext"} set to {false} indicates the end of the _incremental stream_. This
+incremental stream update result is sent when the underlying iterator of the
+`films` field closes.
```
```suggestion
When the underlying iterator of the `films` field closes there is no more data
to deliver, so the third and final _incremental stream update result_ sets
{"hasNext"} to {false} to indicate the end of the _incremental stream_.
```

benjie commented on spec/Appendix E -- Examples.md:
```diff
@@ -0,0 +1,210 @@
+contains the data for the `terrain` field. This _incremental result_ contains a
+{"subPath"} entry to indicate to clients that the _response position_ of this
+result can be determined by concatenating the path from the _pending result_
+with id `"0"` and the value of this {"subPath"} entry.
```
This reads a little ambiguously: one misinterpretation would be "concatenating (the path from the pending result) with (id "0") and (the value of this subpath entry)".

I've tweaked the grammar slightly to avoid this:

```suggestion
result can be determined by concatenating: the path from the _pending result_
for id `"0"`, and the value of this {"subPath"} entry.
```

benjie commented on spec/Appendix E -- Examples.md:
````diff
@@ -0,0 +1,210 @@
+```json example
+{
+  "incremental": [
+    {
+      "id": "0",
+      "data": { "homeWorld": { "name": "Tatooine" } }
+    },
````
I feel like the weirdness of this needs more calling out - name belongs to both id:0 and id:1, so we're kind of using id:0 as a shortcut (but id:1 would also work). We should make it clear to a client that it can't form `HomeWorldFragment` from solely the root data plus the `id:0` stuff and `NameAndHomeWorldFragment` from solely the root data and the `id:1` stuff - they must merge the trees before extracting the data.

robrichard replied:
Added a paragraph here: https://github.com/graphql/graphql-spec/pull/1203/changes#diff-0237c26b29c484df8d43b13dc3a3658867cfd015d8ccea8ed93454c466dbacb3R200-R203

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -70,6 +79,70 @@ The _request error result_ map must not contain an entry with key {"data"}.
+{"hasNext"}. The value of this entry must be {true} if there are any
+_incremental stream update results_ in the _incremental stream_. The value of
+this entry must be {false} if the initial incremental stream result is the last
+response of the incremental stream.
```
Splitting this across two paragraphs is a little weird, I feel like the list of keys up front should be exhaustive.


```suggestion
The _initial incremental stream result_ must contain an entries with keys {"data"} and {"hasNext"}, and may contain entries with keys
{"errors"}, {"pending"}, {"incremental"}, {"completed"}, and {"extensions"}.

The value of {"data"}, {"errors"} and {"extensions"} are defined in the
same way as an _execution result_ as described in the "Data", "Errors", and
"Extensions" sections below.

The value of the {"hasNext"} entry must be {false}
if the initial incremental stream result is the last response of the incremental
stream. Otherwise, {"hasNext"} must be {true}.
```

I'd like to see a non-normative note below this explaining that {"hasNext"} is required only for it to be an initial incremental stream result, and that omitting it is valid if it's the only result, just that it means it's an execution result instead (i.e. there's no "incremental" at all). In this case, the GraphQL response would not return a stream.

This is a terrible version of that note since I'm writing this already over-time on our call...

Note: Omitting {"hasNext"} would make the response a regular _execution result_, which is valid only when returned directly and not as part of a stream.

robrichard replied:
This suggested edit changes pending from required to optional on the initial incremental stream result, I don't think we have a case for an incremental stream without a pending? I will keep it as required when I incorporate this change.

Do you still think we need a note about omitting hasNext? it's a bit more complicated as you also need to omit pending?

I will however add a note explaining the reasoning for allowing hasNext: false in the initial incremental stream result

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -70,6 +79,70 @@ The _request error result_ map must not contain an entry with key {"data"}.
+The _initial incremental stream result_ may contain an entry with they key
+{"incremental"}. The value of this entry must be a non-empty list of
+_incremental result_. Each _incremental result_ must be a map as described in
+the "Incremental Result" section below.
```
I'd add a non-normative note detailing why this is allowed.

benjie replied:
Same for "completed"

robrichard replied:
Note added here: https://github.com/graphql/graphql-spec/pull/1203/changes#diff-98d0cd153b72b63c417ad4238e8cc0d3385691ccbde7f7674bc0d2a718b896ecR116-R126

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -70,6 +79,70 @@ The _request error result_ map must not contain an entry with key {"data"}.
+The _initial incremental stream result_ must contain an entry with the key
+{"pending"}. The value of this entry must be a non-empty list of _pending
+result_. Each _pending result_ must be a map as described in the "Pending
+Result" section below.
```
If "hasNext" is true, pending _must_ be present.
~~If "hasNext" is false, pending _must not_ be present.~~ < Not true, could be hasNext:false along with pending:[...],incremental:[...],completed:[...]

robrichard replied:
Going to leave this as is since `pending` must be present in all cases for the initial incremental stream result

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -70,6 +79,70 @@ The _request error result_ map must not contain an entry with key {"data"}.
+{"pending"}, {"incremental"}, and/or {"completed"}. The value of these entries
+are defined in the same way as an _initial incremental stream result_ as
+described in the "Pending Result", "Incremental Result", and "Completed Result"
+sections below.
```
Should we add that `{"hasNext": true}` alone is invalid?

I propose that we *do not* do so. Not doing so allows `{"hasNext: true}` to be used as a keepalive.

robrichard replied:
Agreed, do you think we should specifically mention this, outside of every other entry being defined as optional?

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -94,6 +167,9 @@ represents a path in the response, not in the request.
 
+When a _response path_ is present on an _incremental result_, it identifies the
+_response position_ of the incremental data update.
+
```
... _can_ a response path be present on an _incremental result_? Maybe you mean _pending result_?

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -323,17 +399,173 @@ discouraged.
+
+Every _pending result_ must contain an entry with the key {"id"} with a string
+value. This {"id"} should be used by clients to correlate pending results with
+_incremental result_ and _completed result_. The {"id"} value must be unique for
```
```suggestion
_incremental result_ and _completed result_. The {"id"} value must be unique across
```

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -323,17 +399,173 @@ discouraged.
+value. This {"id"} should be used by clients to correlate pending results with
+_incremental result_ and _completed result_. The {"id"} value must be unique for
+the entire _incremental stream_ response. There must not be any other pending
+result in the _incremental stream_ that contains the same {"id"}.
```
```suggestion
result in the _incremental stream_ with the same {"id"}.
```

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -323,17 +399,173 @@ discouraged.
+incremental result or completed result. The associated pending result must
+appear in the _incremental stream_, in the same or prior _initial incremental
+stream result_ or _execution update result_ as the given incremental result or
+completed result.
```
```suggestion
:: The _associated pending result_ of an _incremental result_ or _completed
result_ is the _pending result_ whose {"id"} entry has the same value as the
{"id"} entry of the given incremental result or completed result.
```

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -323,17 +399,173 @@ discouraged.
+Every _incremental result_ must contain an entry with the key {"id"} with a
+string value. The definition of _associated pending result_ describes how this
+value is used to determine the associated pending result for a given
+_incremental result_.
```
```suggestion
Every _incremental result_ must contain an entry with the key {"id"}, the value
of which is a string referencing its _associated pending result_. The associated
pending result must appear either in the _initial incremental stream result_,
in a prior _incremental stream updated result_, or in the same _incremental
stream updated result_ as the _incremental result_ that references it.
```

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -323,17 +399,173 @@ discouraged.
+{"items"} and these errors propagate to a _response position_ higher than the
+_incremental list result_'s response position, the incremental list result is
+considered failed and should not be included in the _incremental stream_. The
+errors that caused this failure will be included in a _completed result_.
```
This doesn't feel like the right wording for error propagation here ("response position higher than ..." doesn't feel right) but it's too late at night right now for me to figure out the correct wording.

robrichard replied:
Updating to 
> and these errors propagate to a parent _response position_ of the _incremental list result_'s response position

`parent` and `sibling` are used in the execution section

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -323,17 +399,173 @@ discouraged.
+{"items"} and these errors did not propagate to a path higher than the
+_incremental list result_'s path, the incremental list result must contain an
+entry with key {"errors"} containing these execution errors. The value of this
+entry is described in the "Errors" section.
```
Should we add a non-normative (or normative, maybe) note to say that the response path of these errors will always start with the response position followed by an index? Want to make sure that no-one attempts to attach the error directly to the list, for example.

robrichard replied:
I added more detail to this section that I think makes this more clear.

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -323,17 +399,173 @@ discouraged.
+determined by concatenating the value of the _associated pending result_'s
+{"path"} entry with the value of this {"subPath"} entry. If no {"subPath"} entry
+is present, the _response position_ is the value of the associated pending
+result's {"path"} entry.
```
```suggestion
The _incremental object result_ may contain a {"subPath"} entry. If such an entry
is present, the _response position_ of the incremental object result is the
result of appending the value of this {"subPath"} to the value of the {"path"}
entry of the _associated pending result_. If no {"subPath"} entry is present, the
_response position_ is the value of the associated pending result's {"path"}
entry.
```

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -323,17 +399,173 @@ discouraged.
+result's {"path"} entry.
+
+An _incremental object result_ may be used to deliver data for response fields
+that were contained in more than one deferred fragments. In that case, the
```
```suggestion
that were contained in more than one deferred fragment. In that case, the
```

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -323,17 +399,173 @@ discouraged.
+An _incremental object result_ may be used to deliver data for response fields
+that were contained in more than one deferred fragments. In that case, the
+_associated pending result_ of the incremental object result must be a _pending
+result_ with the longest {"path"}.
```
This is insufficiently precise; it must relate to the deferred fragment. Further, is there a way in which this can be broken when combined with adding incremental/completed to the initial incremental execution result through eager execution and maybe nested fragments? I feel like you might asynchronously complete the requirements and reference a short path, and then a long path comes about later (whilst still executing the non-deferred selection set)...

robrichard replied:
Updated this section to be more precise.

I think since we know about all of the deferred fields after collect fields, the server should always be able to tell which is the longest pending path regardless of the execution order.

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -323,17 +399,173 @@ discouraged.
+_associated pending result_ of the incremental object result must be a _pending
+result_ with the longest {"path"}.
+
+Every _incremental object result_ must contain a {"data"} entry. The {"data"}
+entry must contain a map of additional response fields. The {"data"} entry in an
+incremental object result will be of the type of the field at the incremental
+object result's _response position_.
```
"will be of the type of the field" feels wrong. The field may be `User` but if the selection set is `{ foo: username }` then the "type" is `{ foo: String }`. It's more about the execution of the selections than the type of the field. (Similar concern for lists.)

robrichard replied:
I was basing it off of this existing section, where we say the root data is either the query or mutation root operation type:

> If the operation was a query, this output will be an
object of the query root operation type; if the operation was a mutation, this
output will be an object of the mutation root operation type.

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -323,17 +399,173 @@ discouraged.
+{"data"} and these errors propagated to a _response position_ higher than the
+_incremental object result_'s response position, the incremental object result
+is considered failed and should not be included in the incremental stream. The
+errors that caused this failure will be included in a _completed result_.
```
Is "errors" correct? I think not: when response bubbling causes a selection set to be destroyed you only get one error independent of how many bubbled, and I think that should be the same here?

```suggestion
error that caused this failure will be included in a _completed result_.
```

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -323,17 +399,173 @@ discouraged.
+errors that caused this failure will be included in a _completed result_.
+
+If any _execution error_ were raised during the execution of the results in
+{"data"} and these errors did not propagate to a _response position_ higher than
```
```suggestion
{"data"} and no such error propagated to a _response position_ higher than
```

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -323,17 +399,173 @@ discouraged.
+
+Every _completed result_ must contain an entry with the key {"id"} with a string
+value. The definition of _associated pending result_ describes how this value is
+used to determine the associated pending result for a given _completed result_.
```
```suggestion
Every _completed result_ must contain an entry with the key {"id"}, the
value of which is a string referencing its _associated pending result_. The
associated pending result must appear either in the _initial incremental stream
result_, in a prior _incremental stream updated result_, or in the same
_incremental stream updated result_ as the _completed result_ that references
it.
```

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -323,17 +399,173 @@ discouraged.
+
+Every _incremental list result_ must contain an entry with the key {"id"}, used
+to determine the _associated pending result_ for this _incremental result_.
+
```
Already covered in _incremental result_?

```suggestion
```

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -323,17 +399,173 @@ discouraged.
+_associated pending result_ has failed, due to an execution error propagating to
+a _response position_ higher than the _incremental result_'s response position.
+The {"errors"} entry must contain these execution errors. The value of this
+entry is described in the "Errors" section.
```
🤔  Can errors ever contain more than one error?

Answer: yes - for defer there be many associated selection sets at different levels, and multiple of these may fail due to error propagation.

benjie commented on spec/Section 7 -- Response.md:
```diff
@@ -323,17 +399,173 @@ discouraged.
-clients, the _execution result_ and _request error result_ maps must not contain
-any entries other than those described above. Clients must ignore any entries
-other than those described above.
+clients, any of the maps described in the "Response" section (with the exception
+of {"extensions"}) must not contain any entries other than those described
+above. Clients must ignore any entries other than those described above.
```
We could probably upstream this change.

yaacovCR reviewed (approved):
This is amazing work, and others have given already and I’m sure we’ll continue to give wonderful feedback.

My two cents is that it may be beneficial to refer to the pending/completed and possibly even the incremental entries as notifications rather than results. The result is the deferred/streamed data at that path that has an id. It’s an abstract concept in terms of the response format, to be assembled, perhaps as needed, by the client.

The response can strictly speaking never send a pending result, in fact that’s in some sense an internal contradiction, or, at best, a shorthand for what we really mean, a notification or description of a result that is pending/completed. The incremental entries are actually results/updates, although calling them as such may confuse them with the overarching defer/stream result that they help complete.

I offer this comment just as a suggestion, I could live with this somewhat confusing shorthand if others feel that it is the best compromise in terms of clarity.

robrichard committed:
PR Feedback

robrichard committed:
Pending Result -> Incremental Pending Notice

robrichard committed:
Completed Result -> Incremental Completion Notice

robrichard commented:
@yaacovCR @benjie The following terms have now been updated:

Pending Result => Incremental Pending Notice
Completed Result => Incremental Completion Notice
